src/lib/model/Async2SyncModel.js

```diff
@@ -0,0 +1,272 @@
+ *                                                                        *
+ *  This file is made available under the terms of the license agreement  *
+ *  specified in the corresponding source code repository.                *
+ *                                                                        *
+ *  ORIGINAL AUTHOR:                                                      *
+ *       Paweł Marzec - pawel.marzec@modusbox.com                         *
+ **************************************************************************/
+
+'use strict';
+const util = require('util');
+
+const PSM = require('./common').PersistentStateMachine;
+const MojaloopRequests = require('@mojaloop/sdk-standard-components').MojaloopRequests;
+const deferredJob = require('@internal/shared/deferredJob');
+
+function generate({
+    /**
+     * @name channelNameMethod
+     * @description generates the pub/sub channel name
+     * @param {object} args - the arguments passed as object, same as passed to `run, triggerDeferredJob, generateKey` method
+     * @returns {string} - the pub/sub channel name
+     */
+    channelNameMethod,
+
+    /**
+     * @name requestActionMethod
+     * @description invokes the call to switch
+     * @param {object} requests - MojaloopRequests instance
+     * @param {array} args - the arguments passed as object to `run` method
+     */
+    requestActionMethod,
+
+    /**
+     * @name argsValidationMethod
+     * @description makes validation of args object, invoked in `run, triggerDeferredJob, generateKey` methods to ensure everything is going well
+     * @param {object} requests - MojaloopRequests instance
+     * @param {array} args - the arguments passed as object to `run` method
+     */
+    argsValidationMethod,
+
+    // the name of the model, used for logging
+    modelName
+}) {
+
+
+    const specStateMachine = {
+        init: 'start',
+        transitions: [
+            { name: 'init', from: 'none', to: 'start' },
+            { name: 'requestAction', from: 'start', to: 'succeeded' },
+            { name: 'error', from: '*', to: 'errored' },
+        ],
+        methods: {
+            // workflow methods
+            run,
+            getResponse,
+
+            // specific transitions handlers methods
+            onRequestAction
+        }
+    };
+
+    /**
+     * @name run
+     * @description run the workflow logic
+     * @param {arguments} args - arguments 
+     * @returns {Object} - the http response payload
+     */
+    async function run(args) {
+        // input validation, it should throws if any of args is invalid
+        argsValidationMethod(args);
+
+        const { data, logger } = this.context;
+        try {
+            // run transitions based on incoming state
+            switch (data.currentState) {
+                case 'start':
+                    // the first transition is requestAction
+                    await this.requestAction(args);
```

Comment:
Do you need to pass in the `fsm` here? The function signature below takes 2 arguments.

Reply:
the function below is `onRequestAction` and it is internally invoked by FSM as a result of calling transition `requestAction` here

Reply:
and FSM intentionally pass as first parameter its instance to transition notification handler <- that is the weird way how FSM working (not documented by the way)

Reply:
> @eoln Will you be making additional PRs to refactor all models to this pattern? I have a particular interest in the outbound transfers model; that state machine is a little more complex than the simple "single async to single sync" pattern expressed by your Async2SyncModel.

it is our final goal to be able to specify/generate more complex workflows or at least use some simpler composition pattern. Up to PISP demo (April) time, we(PISP) will treat the rest of the code as technical debt, and refactor when we will be ready -> we need to collect more knowledge, and probably first we need to improve code coverage for these complex workflows before starting refactoring.

Reply:
Many thanks.